### PR TITLE
Westberrytech pr

### PR DESCRIPTION
**Pull request by wb-Joy**
## Description

Added WB32F3G71xx support 

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
This needs to be split up into several PRs.  At the very least, the keyboard code needs to be pulled out into a separate PR, so this is just adding the platform files for the WB32 controller.

**Comment by wb-Joy** [comment]
@drashna  I have deleted the keyboard code , Wait for this pr to pass before submitting the keyboard code.

File: bootloader.mk
Comment (drashna) [inline bootloader.mk:132]: Does this actually support dfu-util?

File: platforms/chibios/drivers/i2c_master.c
Comment (drashna) [inline platforms/chibios/drivers/i2c_master.c:35]: So this is being hard coded to only support this config?  

if so, does this chip support other configurations?  If so, then this should be configurable.

File: platforms/chibios/flash.mk
Comment (drashna) [inline platforms/chibios/flash.mk:27]: does this support dfu-util, rather than a custom/proprietary flasher?

**Review by drashna — CHANGES_REQUESTED** [review]

File: platforms/chibios/flash.mk
Comment (wb-Joy) [reply platforms/chibios/flash.mk:27]: We don't support dfu-util and use our own flasher: DfuFwUpdater_cli.exe , Where should this flasher be submitted ?

File: platforms/chibios/drivers/i2c_master.c
Comment (wb-Joy) [reply platforms/chibios/drivers/i2c_master.c:35]: This is the default configuration, the user can modify the configuration by redefining WB32_I2CCONFIG

File: bootloader.mk
Comment (wb-Joy) [reply bootloader.mk:132]: dfu-util is not supported, can I delete this code directly ?

File: platforms/chibios/flash.mk
Comment (fauxpark) [reply platforms/chibios/flash.mk:27]: Is this tool open source?

Comment (wb-Joy) [reply platforms/chibios/flash.mk:27]: We will open source and will upload to github today.

File: bootloader.mk
Comment (fauxpark) [reply bootloader.mk]: The dfu-util related sutff can be deleted but it would be useful to keep the `wb32-dfu` bootloader option.

File: platforms/chibios/flash.mk
Comment (drashna) [reply platforms/chibios/flash.mk:36]: That appears to be windows only?

Comment (wb-Joy) [reply platforms/chibios/flash.mk:36]: @drashna  Yea, currently only windows are supported.

Comment (fauxpark) [reply platforms/chibios/flash.mk:36]: So what can people do if they want to flash their board from a Linux or macOS machine?

Comment (wb-Joy) [reply platforms/chibios/flash.mk:36]: @fauxpark Currently, we hope users to use Windows to flash their board.

File: platforms/chibios/drivers/i2c_master.c
Comment (drashna) [reply platforms/chibios/drivers/i2c_master.c:35]: Ideally, would want to have the config set up like the rest of it, so that other people using the same or similar chips don't have to add the full config, just add defines based on their usage.

File: platforms/chibios/flash.mk
Comment (drashna) [reply platforms/chibios/flash.mk:36]: > @fauxpark Currently, we hope users to use Windows to flash their board.

As somebody that doesn't use Windows for their desktop machine anymore, this leaves a sour taste in my mouth. 

Since the whole of QMK supports Linux, MacOS and Windows, adding a windows only tool doesn't feel right.  And you are dumping the work on the community and QMK to create a linux and macOS utility.

**Review by drashna — DISMISSED** [review]

**Review by wb-Joy — COMMENTED** [review]
pass

**Comment by wb-Joy** [comment]
@drashna 
Excuse me，How to solve this error?
图片

**Comment by drashna** [comment]
should be fine to ignore this specific error, IIRC.

**Comment by wb-Joy** [comment]
@drashna  May I ask when this PR can be merged or is there anything else that needs to be fixed? Our company is in a hurry.

**Comment by drashna** [comment]
It's waiting on another approval, as we generally only merge PRs after 2 approvals, especially for core changes.

However, the hope is that this gets into this breaking changes cycle, so that it's merged into master, at the end of the month.

**Comment by wb-Joy** [comment]
@drashna Can you help me to urge the another approver? He never replied to me. Please

File: platforms/chibios/drivers/uart.c
Comment (zvecr) [inline platforms/chibios/drivers/uart.c:45]: Should be reverted as reintroducing the STM32_defines is wrong.

Instead `PAL_WB32_PUPDR_PULLUP` should be defined to `PAL_OUTPUT_TYPE_OPENDRAIN` within `tmk_core/common/chibios/chibios_config.h` per the other compatibility defines.

File: platforms/chibios/drivers/spi_master.c
Comment (zvecr) [inline platforms/chibios/drivers/spi_master.c:54]: Revert + implement same change as uart

File: platforms/chibios/drivers/i2c_master.c
Comment (zvecr) [inline platforms/chibios/drivers/i2c_master.c:84]: Revert + implement same change as uart

**Review by zvecr — CHANGES_REQUESTED** [review]

File: lib/python/qmk/cli/doctor/linux.py
Comment (zvecr) [inline lib/python/qmk/cli/doctor/linux.py:65]: Adding this here without updating the udev rules template is wrong, as it will constantly report issues.

File: platforms/chibios/drivers/spi_master.c
Comment (zvecr) [inline platforms/chibios/drivers/spi_master.c:58]: Would be nice to avoid a different style of implementation than the HT32/K20x ifdefs below.

File: tmk_core/common/chibios/bootloader.c
Comment (zvecr) [inline tmk_core/common/chibios/bootloader.c:98]: Not a fan of the duplication for WB32_BOOTLOADER_ADDRESS, as the code is virtually the same.

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (zvecr) [inline tmk_core/common/chibios/eeprom_wb32.c:1]: Does this add any value outside implementing a flash abstraction and reusing the stm wear leveling code?

**Review by zvecr — CHANGES_REQUESTED** [review]

File: builddefs/mcu_selection.mk
Comment (zvecr) [inline builddefs/mcu_selection.mk:700]:
```suggestion
```

**Comment by wb-Joy** [comment]
@zvecr  I have solved the problem you raised. Please check whether there are other problems.

File: platforms/chibios/drivers/spi_master.c
Comment (wb-Joy) [reply platforms/chibios/drivers/spi_master.c]: Resolved

File: platforms/chibios/drivers/i2c_master.c
Comment (wb-Joy) [reply platforms/chibios/drivers/i2c_master.c]: Resolved

File: platforms/chibios/drivers/spi_master.c
Comment (wb-Joy) [reply platforms/chibios/drivers/spi_master.c]: Resolved

File: tmk_core/common/chibios/bootloader.c
Comment (wb-Joy) [reply tmk_core/common/chibios/bootloader.c]: Resolved

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (wb-Joy) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: Resolved

Comment (zvecr) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: > Resolved

Not really, as you still not given justification to why it cannot use the existing wear leveling code or swapped it over to use said code.

Comment (wb-Joy) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: Because the flash operation of the WB32 MCU is different from other MCU operations

Comment (zvecr) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: "different" isnt really answering why it cant reuse the existing code within `eeprom_stm` and implement its own `flash_wb` logic.

From `EEPROM_WriteDataByte`
> If a byte is already written, the whole page must be copied to a buffer, the byte changed and the manipulated buffer written after PageErase.

```c
    WBFLASH_Read(pagpos * FEE_PAGE_SIZE + WB32_FLASH_BASE, wb_flash_buffer, FEE_PAGE_SIZE);
    wb_flash_buffer[pagoff] = DataByte;
    /* Erase the specified FLASH page */
    FMC_ErasePage(pagpos * FEE_PAGE_SIZE + WB32_FLASH_BASE);
```

Thats not what this code actually does, however the implications are actually worse. The flash wear cylcles are unnecessarily burnt. Using existing code where possible ensures issues like this are potentially mitigated.

Comment (tzarc) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: As per the datasheet the flash has 100k write cycle endurance.
If writing a single effective byte to the flash page requires reading 256 bytes, then writing 256 bytes, the endurance may potentially be exceeded within a few weeks, potentially corrupting any stored data and/or preventing any future data storage.

Suggest re-thinking the decision to not integrate with the existing wear-leveling code, which erases pages only when required.

Comment (tzarc) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: In fact, in the reference manual, section 2.4:

* The flash should be program by page (256 byte), byte program operation is not supported.
* The flash page should be erased before program, it is not recommend to program the same page multiple times without erase.

Can you clarify the second point? Given these two statements as they stand, I'm inclined to reject the inclusion of this emulated EEPROM implementation entirely, given that the likelihood of failure is going to be very high in a short period of time.

**Review by wb-Joy — COMMENTED** [review]
WB32 flash has 200k+ write cycle endurance. 
I have modified the code, please check whether there are other problems.

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (wb-Joy) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: WB32 flash has 200k+ write cycle endurance.
I have modified the code, please check whether there are other problems.

Comment (wb-Joy) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: If it doesn't pass, can I delete this part of the code and pass?

Comment (zvecr) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: > Can you clarify the second point?

Can you actually answer rather than marking this as resolved.

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (zvecr) [inline tmk_core/common/chibios/eeprom_wb32.c:229]: > If a byte is already written,

The function docs **still** dont match the behaviour. All that has been added is an optimisation when writes align to page size, along with a load of duplication.

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (tzarc) [reply tmk_core/common/chibios/eeprom_wb32.c:1]: We can see that from the datasheet.

The issue is that if we go ahead with this current implementation, the effective lifetime of the pages used will be severely diminished, given that the number of write cycles for the same byte locations is limited to 200k (despite the datasheet saying 100k).

Every write to every location results in the read/erase/write of a whole page of 256 bytes. Given that the vast majority of core QMK EEPROM usage occurs in the first 35 bytes, the first page may be rendered unusable within a few weeks, depending on whether the user regularly modifies their RGB settings for instance.

If we allow this implementation to go into the repository, any keyboard manufacturer using them will need to be informed that they may see high failure rates.

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (tzarc) [inline tmk_core/common/chibios/eeprom_wb32.c:215]: Also this comment seems to imply that the page size is 512 bytes.

Comment (wb-Joy) [reply tmk_core/common/chibios/eeprom_wb32.c]: It's my fault, it's been modified.

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (wb-Joy) [reply tmk_core/common/chibios/eeprom_wb32.c]: I'll remove the code related to the eeprom_wb32 first.

**Comment by wb-Joy** [comment]
I'll remove the code related to the eeprom_wb32 first.
Our customers can use external eeprom to store data.

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (wb-Joy) [reply tmk_core/common/chibios/eeprom_wb32.c]: > 
> 
> > Can you clarify the second point?
> 
> Can you actually answer rather than marking this as resolved.

This is determined by our Flash technical craft.

**Comment by peepeetee** [comment]
Is there any language help I can offer here? I could be reached at peepeetee@protonmail.com
在？我来帮您说英语？在peepeetee@protonmail.com联系我

File: tmk_core/common/chibios/eeprom_wb32.c
Comment (peepeetee) [reply tmk_core/common/chibios/eeprom_wb32.c]: @wb-Joy has corresponded with me and wanted me to state that they:
want the see this PR merged. To achieve that they want to not submit this part of the code for now, and submit this at a later time after this breaking change as an update.

>Flash technical craft
by this they mean flash fabrication technology

**Comment by drashna** [comment]
This has some merge conflicts that need to be resolved.

**Comment by wb-Joy** [comment]
> This has some merge conflicts that need to be resolved.

I have solved it, please review it again.

**Review by tzarc — APPROVED** [review]

**Comment by wb-Joy** [comment]
Is it ok to pass now?
Two reviewers have been approved.

**Comment by tzarc** [comment]
> Is it ok to pass now? Two reviewers have been approved.

There's only one.

image

**Comment by wb-Joy** [comment]
I see it.

**Comment by wb-Joy** [comment]
> > Is it ok to pass now? Two reviewers have been approved.
> 
> There's only one green tick.
> 
> image

I see it.
02F0EEEF-F439-484E-87A3-79057811D438

**Comment by tzarc** [comment]
That was old and has since been rescinded.

image

**Comment by wb-Joy** [comment]
Okay, so now I just need to wait other reviewer approval ？

**Comment by tzarc** [comment]
Yes.

**Comment by wb-Joy** [comment]
Thanks ！

**Comment by wb-Joy** [comment]
@zvecr Could you please review this PR？

**Review by zvecr — APPROVED** [review]

**Pull request merged**
